npep/npep-126-egress-traffic-control.md

```diff
@@ -0,0 +1,93 @@
+# NPEP-126: Add northbound traffic support in (B)ANP API
+
+* Issue: [#126](https://github.com/kubernetes-sigs/network-policy-api/issues/126)
+* Status: Provisional
+
+## TLDR
+
+This NPEP proposes adding support for cluster egress (northbound) traffic control in the `AdminNetworkPolicy` and `BaselineAdminNetworkPolicy` API objects.
+
+## Goals
+
+* Implement egress traffic control towards external destinations (outside the cluster)
+* Implement egress traffic control towards cluster nodes
+  - Currently the behaviour for policies defined around traffic from cluster
+    workloads (non-hostNetworked pods) towards nodes in the
+    cluster is undefined. See https://github.com/kubernetes-sigs/network-policy-api/issues/73.
+    - ANP currently supports only east-west traffic and this traffic flow cuts from
+    overlay to underlay which makes this part of the egress (northbound) use case.
+    - Let's provide a defined behaviour in ANP to explicitly achieve the use case.
+    - NOTE: Traffic towards nodes here includes traffic towards host-networked pods on that node
+      because a "node" resource encompasses all objects that share the host-networking resources
+* Implement egress traffic control towards hostNetworked pods
+  - Currently the behaviour for policies defined around traffic from cluster
+  workloads (non-hostNetworked pods) towards hostNetworked pods in the
+  cluster is undefined. See https://github.com/kubernetes-sigs/network-policy-api/issues/73.
+  - ANP currently supports only east-west traffic and this traffic flow cuts from
+  overlay to underlay which makes this part of the egress (northbound) use case.
+  - Let's provide a defined behaviour in ANP to explicitly achieve the use case.
+  - NOTE: We can keep this goal experimental in case implementors cannot distinguish between
+    a node and host-networked pod
+
+## Non-Goals
+
+* Implementing southbound (ingress) traffic use cases is outside the scope of this NPEP
+
+## Introduction
+
+### User Stories for egress traffic control towards external destinations
+
+1. **As a** cluster administrator **I want** to restrict traffic from
+specific cluster workloads to all or specific destinations outside the
+cluster **so that** I can enforce security for northbound traffic.
+Example: Pods in namespaceA and namespaceB should not be able to talk
+to the internet but they should be able to access company's intranet.
+
+2. **As a** cluster administrator **I want** to explicitly allow traffic
+from cluster workloads to handpicked destinations outside the cluster and
+deny the rest **so that** I can guarantee cluster wide egress traffic
+flow irrespective of whatever network policies are defined by the
+developer for specific namespaces. Example: As an owner of namespaceA I
+define policies that deny all northbound egress traffic for that
+namespace. However the cluster-admin can decide all namespaces in the
+cluster must be able to talk to the EXTERNAL_DNS_SERVER_IP on port 53.
```

Comment:
"and deny the rest" (at the top) is unnecessary and confusing for this use case; the problem here is that the namespace admin is denying egress to the DNS server, and the cluster admin wants to undo that. It doesn't matter for that story whether the cluster admin is also denying access to some things.

And you can just turn the "Example" into the user story. User stories should be specific and evocative anyway, so readers can understand exactly what's going on and extrapolate. So, "As a cluster administrator, I want to ensure that pods can reach my cluster-external DNS server even if namespace admins create NetworkPolicies that block cluster-external egress."

Reply:
> "and deny the rest" (at the top) is unnecessary and confusing for this use case; the problem here is that the namespace admin is denying egress to the DNS server, and the cluster admin wants to undo that. It doesn't matter for that story whether the cluster admin is also denying access to some things.

true, removing that bit.

> 
> And you can just turn the "Example" into the user story. User stories should be specific and evocative anyway, so readers can understand exactly what's going on and extrapolate. So, "As a cluster administrator, I want to ensure that pods can reach my cluster-external DNS server even if namespace admins create NetworkPolicies that block cluster-external egress."

ack changing this use story into what you have described :) because it makes sense.